File: skills/motion-world/references/runtimes/react-native/motionProgress.ts
export const clampProgress = (value: number): number =>
  Number.isFinite(value) ? Math.min(1, Math.max(0, value)) : 0;

export const elapsedProgress = (nowMs: number, startMs: number, endMs: number): number => {
  const duration = endMs - startMs;
  if (duration <= 0) return nowMs >= endMs ? 1 : 0;
  return clampProgress((nowMs - startMs) / duration);
};

export const countdownProgress = (remaining: number, total: number): number => {
  if (total <= 0) return remaining <= 0 ? 1 : 0;
  return clampProgress(1 - remaining / total);
};

export const countProgress = (current: number, goal: number, minimum = 0): number => {
  const range = goal - minimum;
  if (range <= 0) return current >= goal ? 1 : 0;
  return clampProgress((current - minimum) / range);
};

File: skills/motion-world/references/runtimes/android/MotionProgress.kt
package motionworld

import kotlin.math.max
import kotlin.math.min

object MotionProgress {
    fun clamp(value: Float): Float = if (value.isFinite()) min(1f, max(0f, value)) else 0f

    fun elapsed(nowMillis: Long, startMillis: Long, endMillis: Long): Float {
        val duration = endMillis - startMillis
        if (duration <= 0L) return if (nowMillis >= endMillis) 1f else 0f
        return clamp((nowMillis - startMillis).toFloat() / duration.toFloat())
    }

    fun countdown(remaining: Float, total: Float): Float {
        if (total <= 0f) return if (remaining <= 0f) 1f else 0f
        return clamp(1f - remaining / total)
    }

    fun count(current: Float, minimum: Float = 0f, goal: Float): Float {
        val range = goal - minimum
        if (range <= 0f) return if (current >= goal) 1f else 0f
        return clamp((current - minimum) / range)
    }
}
